File: app/src/main/java/org/mifos/mobilebanking/presenters/base/BasePresenter.java
package org.mifos.mobilebanking.presenters.base;

import android.content.Context;

import org.mifos.mobilebanking.ui.views.base.MVPView;

/**
 * @author ishan
 * @since 19/05/16
 */
public class BasePresenter<T extends MVPView> implements Presenter<T> {

    protected Context context;
    private T mMvpView;

    protected BasePresenter(Context context) {
        this.context = context;
    }

    @Override
    public void attachView(T mvpView) {
        mMvpView = mvpView;
    }

    @Override
    public void detachView() {
        mMvpView = null;
    }

    public boolean isViewAttached() {
        return mMvpView != null;
    }

    public T getMvpView() {
        return mMvpView;
    }

    public void checkViewAttached() {
        if (!isViewAttached()) throw new MvpViewNotAttachedException();
    }

    public static class MvpViewNotAttachedException extends RuntimeException {
        public MvpViewNotAttachedException() {
            super("Please call Presenter.attachView(MvpView) before" +
                    " requesting data to the Presenter");
        }
    }

}

File: app/src/main/java/org/mifos/mobile/presenters/base/BasePresenter.kt
package org.mifos.mobile.presenters.base

import android.content.Context

import org.mifos.mobile.ui.views.base.MVPView

/**
 * @author ishan
 * @since 19/05/16
 */
open class BasePresenter<T : MVPView?> protected constructor(protected var context: Context?) : Presenter<T> {
    var mvpView: T? = null
        private set

    override fun attachView(mvpView: T) {
        this.mvpView = mvpView
    }

    override fun detachView() {
        mvpView = null
    }

    val isViewAttached: Boolean
        get() = mvpView != null

    fun checkViewAttached() {
        if (!isViewAttached) throw MvpViewNotAttachedException()
    }

    class MvpViewNotAttachedException : RuntimeException("Please call Presenter.attachView(MvpView) before" +
            " requesting data to the Presenter")
}